safe_url_check/lib/src/safe_url_check.dart

```diff
@@ -0,0 +1,144 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:io';
+
+import 'package:retry/retry.dart';
+
+import 'private_ip.dart';
+import 'unique_local_ip.dart';
+import 'version.dart';
+
+const defaultUserAgent = 'package:safe_url_check/$packageVersion '
+    '(+https://github.com/google/dart-neats/tree/master/safe_url_check)';
+
+Future<bool> doSafeUrlCheck(
+  Uri url, {
+  int maxRedirects = 8,
+  String userAgent = defaultUserAgent,
+  HttpClient? client,
+  RetryOptions retryOptions = const RetryOptions(maxAttempts: 3),
+  Duration timeout = const Duration(seconds: 90),
+}) async {
+  ArgumentError.checkNotNull(url, 'url');
+  ArgumentError.checkNotNull(maxRedirects, 'maxRedirects');
+  ArgumentError.checkNotNull(userAgent, 'userAgent');
+  ArgumentError.checkNotNull(retryOptions, 'retryOptions');
+  if (maxRedirects < 0) {
+    throw ArgumentError.value(
+      maxRedirects,
+      'maxRedirects',
+      'must be a positive integer',
+    );
+  }
+
+  try {
+    // Create client if one wasn't given.
+    var c = client;
+    c ??= HttpClient();
+    try {
+      return await SafeUrlChecker(
+        client: c,
+        userAgent: userAgent,
+        retryOptions: retryOptions,
+        timeout: timeout,
+      ).checkUrl(url, maxRedirects);
+    } finally {
+      // Close client, if it was created here.
+      if (client == null) {
+        c.close(force: true);
+      }
+    }
+  } on Exception {
+    return false;
+  }
+}
+
+class SafeUrlChecker {
```

Comment:
Why are we creating a class that has no purpose?

This is just a convoluted way to make function, right?

Why not just keep the old `_safeUrlCheck` make it `safeUrlCheck` in the `safe_url_check/lib/src/safe_url_check.dart` file and add a parameter `skipLocalNetworkCheck` which defaults to `false` and has `@visibleForTestingOnly` annotation.

Or we could just accept that we don't have testing, or use httpbin.org for testing.

Reply:
Or put the parameter on `doSafeUrlCheck`.

This introduces a class `SafeUrlChecker` which we only every construct inorder to call a single method.

It's just a confusing way to have a function that you can pass arguments to in two different ways.

Another, somewhat simpler and ugly option is to make a global:
```dart
@visibleForTestingOnly
bool skipLocalNetworkCheck = false;
```

Put this in an `lib/src/...` file and check it from within `safeUrlCheck`.
If no code actively sets this property, hopefully the compiler will optimize it away.

Reply:
> This is just a convoluted way to make function, right?

I didn't like the recursive function with 6+ parameters. The class keeps the non-changing parts as fields.

Reply:
Updated, PTAL.